File: src/main/java/io/github/colintimbarndt/chat_emotes/data/unicode/joiner/RangeTransformer.java
package io.github.colintimbarndt.chat_emotes.data.unicode.joiner;

import io.github.colintimbarndt.chat_emotes.data.unicode.pattern.MatchResult;
import org.jetbrains.annotations.NotNull;

public final class RangeTransformer implements MatchTransformer {
    private final int base;
    private final int width;
    private final boolean reversed;

    public RangeTransformer(int from, int to) {
        if (from <= to) {
            this.base = from;
            this.width = 1 + to - from;
            this.reversed = false;
        } else {
            this.base = to;
            this.width = 1 + from - to;
            this.reversed = true;
        }
    }

    @Override
    public void transform(@NotNull MatchResult r, @NotNull StringBuilder result) {
        final var ch = reversed ? base - r.index() : base + r.index();
        if (Character.isBmpCodePoint(ch)) {
            result.append((char) ch);
        } else {
            result.append(Character.highSurrogate(ch))
                    .append(Character.lowSurrogate(ch));
        }
    }

    @Override
    public int acceptedWidth() {
        return width;
    }
}

File: common/src/main/kotlin/io/github/colintimbarndt/chat_emotes/common/data/unicode/joiner/RangeTransformer.kt
package io.github.colintimbarndt.chat_emotes.common.data.unicode.joiner

import io.github.colintimbarndt.chat_emotes.common.data.unicode.pattern.MatchResult

class RangeTransformer(from: Int, to: Int) : MatchTransformer {
    private var base = 0
    override val acceptedWidth: Int
    private var reversed = false

    init {
        if (from <= to) {
            base = from
            acceptedWidth = 1 + to - from
            reversed = false
        } else {
            base = to
            acceptedWidth = 1 + from - to
            reversed = true
        }
    }

    override fun transform(r: MatchResult, result: StringBuilder) {
        val ch = if (reversed) base - r.index else base + r.index
        result.appendCodePoint(ch)
    }
}